Release2.0

lnagle opened the pull request.

lnagle committed:
DDQ version 1.1

lnagle committed:
Set isListening flag on close

lnagle committed:
Fix Tests

lnagle committed:
Update backend-mock version

lnagle committed:
Code Review Changes

lnagle committed:
Add comma

lnagle committed:
Get coverage back up to 100%

lnagle committed:
Changes based on code review

lnagle committed:
Code review changes

lnagle committed:
Add manual testing config to gitignore

lnagle committed:
Clear cache

lnagle committed:
Cleared cache

lnagle committed:
Begin manual testing

lnagle committed:
In progress testing framework

lnagle committed:
Write functional tests. In progress

lnagle committed:
Make some necessary changes to DDQ and add more functional tests

quantumew committed:
Fix some tests and remove debug code.

quantumew committed:
Remove obselete tests.

quantumew committed:
Fixing some tests and removing mysql stuff.

quantumew committed:
Removing test scripts from travis. Just want to get it to pass.

quantumew committed:
Finish catching the edge cases.

quantumew committed:
Adding all contributors.

lnagle committed:
Add functional tests, bug fixes, etc

lnagle committed:
Revert testing config and mysql-commands to what the plugins will use during builds

lnagle committed:
Remove functional testing script (It's in ddqMysql now)

lnagle committed:
Remove func* from test script

lnagle committed:
Merge in Andy'
s testing changes

lnagle committed:
Fix broken tests

AbsentSemicolon commented on lib/config-validation.js:
```diff
@@ -0,0 +1,37 @@
+
+module.exports = () => {
+    /**
+     * Checks that the config has all necessary properties to work with the
```
Don't need that.

lnagle replied:
Fixed.

AbsentSemicolon commented on lib/ddq.js:
```diff
@@ -95,18 +91,42 @@ module.exports = (EventEmitter, timers) => {
+
+
+    /**
+     * Calls the backend to stop listening and sets the isListening flag.
```
It looks like it unsets the listening flag since it sets it to false.

lnagle replied:
Fixed.

AbsentSemicolon commented on lib/ddq.js:
```diff
@@ -118,32 +138,72 @@ module.exports = (EventEmitter, timers) => {
 
     /**
-     * Class for DDQ.
+     * Signals that a message is done being processed.
```
Don't need that.

lnagle replied:
Fixed.

AbsentSemicolon commented on lib/ddq.js:
```diff
@@ -118,32 +138,72 @@ module.exports = (EventEmitter, timers) => {
+
+
+    /**
+     * Checks that a callback was passed as the first argument. If it
```
Don't need that.

lnagle replied:
Fixed.

AbsentSemicolon commented on lib/ddq.js:
```diff
@@ -118,32 +138,72 @@ module.exports = (EventEmitter, timers) => {
+
+
+    /**
+     * DDQ class.
```
We could tag this using jscode.

http://usejsdoc.org/tags-class.html

Can also say it extends (augments) something: http://usejsdoc.org/tags-augments.html

lnagle replied:
Seems only beneficial if you're using a function as a class. The class declaration is two lines down. Leaving as is.

AbsentSemicolon replied:
Fair enough, I'm good with that.

lnagle93 commented on lib/ddq.js:
```diff
@@ -118,32 +138,72 @@ module.exports = (EventEmitter, timers) => {
+
+
+    /**
+     * DDQ class.
```
Added.

AbsentSemicolon commented on script/send-message-test.js:
```diff
@@ -0,0 +1,73 @@
+        }
+
+        console.log("Connection was successfully made");
+        fnCallTime = Date.now() / 1000;
```
Maybe use utc date/time here to stay consistent.

lnagle replied:
This was actually slated to be removed. Fixed.

AbsentSemicolon commented on script/send-message-test.js:
```diff
@@ -0,0 +1,73 @@
+                return;
+            }
+
+            console.log("sendMessage completed. Time difference between time of call and completion:", Date.now() / 1000 - fnCallTime);
```
same here, use utc date/time.

lnagle replied:
This was actually slated to be removed. Fixed.

AbsentSemicolon commented on script/wrapped-message-tests.js:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * Inserts a record into the database using the provided query. On error, the
+ * database the database is wiped and the error is handled. If successful, the
```
you have "the database" twice.

lnagle replied:
Fixed.

AbsentSemicolon commented on spec/lib/ddq.spec.js:
```diff
@@ -1,20 +1,23 @@
 "use strict";
 
 describe("tests", () => {
```
Should this be DDQ instead of tests?

lnagle replied:
Correct. Fixed.

lnagle committed:
Chris' Code Review Changes

AbsentSemicolon commented:
Looks good, approved, just need tests to work.

fidian commented on lib/ddq.js:
```diff
@@ -79,14 +75,14 @@ module.exports = (EventEmitter, timers) => {
             wrappedMessage.heartbeat((err) => {
                 if (err) {
-                    ddqInstance.emit(EMIT_ERROR, err);
+                    ddqBackendInstance.emit(EMIT_ERROR, err);
```
The backend instance should not be emitting in this fashion.  If there's a problem with the heartbeat, the main DDQ instance should be emitting.

This works because you have code later that will relay all errors from the backend through the DDQ instance.  My point is that it should not originate from the backend and simply be emitted from the DDQ instance.

quantumew replied:
If you look at where this is called is actually correct, the name of the argument and JSDoc was wrong. I fixed that 7fa4cd34c7465086fb0d53c3f159322448d968d7

fidian commented on lib/ddq.js:
```diff
@@ -152,101 +212,149 @@ module.exports = (EventEmitter, timers) => {
-            if (!callback) {
-                callback = noop;
-            }
+        sendMessage(callback, message, topic) {
```
Callback should be last for all functions, here and in the plugin.

`validateArgs()`, I believe, should be eliminated and replaced with a simple `if` statement or even a conditional.

    callback = callback || noop;

quantumew replied:
7fa4cd34c7465086fb0d53c3f159322448d968d7

fidian commented on lib/ddq.js:
```diff
@@ -152,101 +212,149 @@ module.exports = (EventEmitter, timers) => {
+                args[0] = applyCloseConditions(this, args[0]);
+
+                // Topic must be a non-empty string.
+                if (!args[2] || typeof args[2] !== "string" || !args[2].length) {
```
Why must topic be a non-empty string?
Is that a concern of DDQ or of the backend?

quantumew replied:
Nope, in fact the backend plugin already deals with this. 7fa4cd34c7465086fb0d53c3f159322448d968d7

fidian commented on mysql-commands.sql:
```diff
@@ -0,0 +1,14 @@
+DROP DATABASE IF EXISTS testQueue;
```
Should this file exist?  Shouldn't the DDQ classes be exercised through the mock?

quantumew replied:
Apparently the DDQ backend MySQL plugin uses this for testing. Not sure why it is in this repository and not in the backend plugin repository. I think he was thinking other plugins in the future could use it.

fidian replied:
That seems wrong to me.  It's not a shared asset and can only work for MySQL.

quantumew replied:
I agree. Let me move it to plugin.

fidian commented on package.json:
```diff
@@ -1,20 +1,36 @@
         "codecov": "^1.0.1",
-        "ddq-backend-mock": "^1.0.0",
+        "ddq-backend-mock": "git://github.com/tests-always-included/ddq-backend-mock.git#release2.0",
+        "ddq-backend-mysql": "git://github.com/tests-always-included/ddq-backend-mysql.git#release2.0",
```
I think this should not be used.  Also, it would be nice to not point to a git repository and instead publish 2.0 of the mock.

quantumew replied:
7fa4cd34c7465086fb0d53c3f159322448d968d7

fidian commented on package.json:
```diff
@@ -30,8 +46,8 @@
         "watch": "jasmine-node node_modules/jasmine-test-helpers/lib/ lib spec --autotest --watch"
     },
-    "version": "1.0.0"
+    "version": "1.1.0"
```
Are you releasing 1.1.0 or 2.0.0?

quantumew replied:
Bumped version to 2 7fa4cd34c7465086fb0d53c3f159322448d968d7

quantumew committed:
Code review fixes.

Renamed improperly names argument, bump version, move callback to last
argument, and update tests.

fidian commented on lib/ddq.js:
```diff
@@ -331,29 +312,21 @@ module.exports = (configValidation, EventEmitter, timers) => {
-
-            args = validateArgs(callback, message, topic);
+        sendMessage(message, topic, callback) {
+            callback = callback || noop;
```
Topic is optional, so `sendMessage(message, callback)` will fail unless you add this code.

    if (typeof message === "function") {
        callback = topic;
        topic = null;
    }

quantumew replied:
Oh right. 9398f3a7a30533f9ef80ce7a58f0e1b9b0ef9546

fidian commented on spec/lib/ddq.spec.js:
```diff
@@ -240,10 +240,10 @@ describe("DDQ", () => {
             ddq.open();
             ddq.isBusy = true;
-            ddq.sendMessage((err) => {
+            ddq.sendMessage("message", "topic", (err) => {
```
Probably missing a test where the topic is omitted.

quantumew replied:
Yup after adding the above code this test case was missing 9398f3a7a30533f9ef80ce7a58f0e1b9b0ef9546

fidian reviewed (changes requested):
Looks good, but a few more changes please.

quantumew committed:
Moving mysql specific stuff and allowing topic to be optional.

quantumew committed:
Fix test. Was calling done twice

quantumew commented:
Fixed duplicate call to `done` @fidian 429aef263d849d5c72b508df5b7b6552e98252c3

fidian commented:
Awesome.  Merging!  Thanks @lnagle93 and @quantumew!
